### PR TITLE
add smape loss function

### PR DESCRIPTION
**Pull request by lakshayg**
Refer #203

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en.

**NOTE**: I was unable to build the master branch on my machine and so this patch is also NOT tested and may require careful review.

### CODE REVIEW

**Comment by lakshayg** [comment]
Can one of the devs please look into the appveyor failure, I am not sure why this fails.

**Comment by annaveronika** [comment]
Sorry for a delay, we had long holidays. This check can be ignored, we'll review the code soon.

File: catboost/libs/options/catboost_options.cpp [hunk 1]
```diff
@@ -259,7 +266,8 @@ namespace NCatboostOptions {
         if (lossFunction == ELossFunction::Quantile ||
             lossFunction == ELossFunction::MAE ||
             lossFunction == ELossFunction::LogLinQuantile ||
-            lossFunction == ELossFunction::MAPE)
+            lossFunction == ELossFunction::MAPE ||
+            lossFunction == ELossFunction::SMAPE)
```
Comment (Noxoomo) [inline catboost/libs/options/catboost_options.cpp]: SMAPE defined with second order derivates, but here you ensure, that you never uses them
If loss has 2-order ders, then we can use Newton in leaves estimation

**Comment by Noxoomo** [comment]
Hi!

Do you know any widely-used freely-available dataset where one need  to optimise  this loss-function? We would like to add it to our tests as well as to tune default params for this loss

File: catboost/libs/algo/error_functions.h [hunk 2]
```diff
@@ -295,6 +295,29 @@ class TMAPError : public IDerCalcer<TMAPError, /*StoreExpApproxParam*/ false> {
...
 };
 
+class TSMAPError : public IDerCalcer<TSMAPError, /*StoreExpApproxParam*/ false> {
+public:
+    explicit TSMAPError(bool storeExpApprox) {
+        CB_ENSURE(storeExpApprox == StoreExpApprox, "Approx format does not match");
+    }
+
+    double CalcDer(double approx, float target) const {
+        const float sumabs = abs(target) + abs(approx);
```
Comment (Noxoomo) [inline catboost/libs/algo/error_functions.h]: CPU computations should be in double

Comment (lakshayg) [reply catboost/libs/algo/error_functions.h]: Done!

**Comment by lakshayg** [comment]
@Noxoomo I am not familiar with any widely-used freely-available dataset but I have seen people using this loss function on kaggle [Kaggle Web Traffic Competition][1].

[1]: https://www.kaggle.com/c/web-traffic-time-series-forecasting/discussion

File: catboost/libs/options/catboost_options.cpp (hunk 1 above)
Comment (lakshayg) [reply catboost/libs/options/catboost_options.cpp]: Done!

**Comment by annaveronika** [comment]
@lakshayg ping

**Comment by lakshayg** [comment]
Hi, completely forget about this. Will make the changes this weekend.
Thanks for reminding :)

**Comment by lakshayg** [comment]
@annaveronika  Could you or someone else look into the failures. I am not sure what is causing them.

**Comment by annaveronika** [comment]
@lakshayg You can ignore this error for now. We will look into it later, it is not related to your pr

**Comment by annaveronika** [comment]
@lakshayg you've added an SMAPE test in test.py. The test fails, could you make a fix?
The test runs 
./catboost fit --loss-function SMAPE -f ../pytest/data/adult/train_small  -t ../pytest/data/adult/test_small --cd ../pytest/data/adult/train.cd -i 10 -T 4 -r 0

**Comment by annaveronika** [comment]
I reverted for now, because test of SMAPE is failing.
To run the test you need to run
>ya make -t -A
in directory catboost/pytest

To canonize the test results you need to run
>ya make -t -A -Z

File: catboost/app/bind_options.cpp [hunk 3]
```diff
@@ -108,14 +108,14 @@ void ParseCommandLine(int argc, const char* argv[],
     BindPoolLoadParams(&parser, params);
 
     parser.AddLongOption("loss-function",
-                         "Should be one of: Logloss, CrossEntropy, RMSE, MAE, Quantile, LogLinQuantile, MAPE, Poisson, MultiClass, MultiClassOneVsAll, PairLogit, QueryRMSE, QuerySoftMax. A loss might have params, then params should be written in format Loss:paramName=value.")
+                         "Should be one of: Logloss, CrossEntropy, RMSE, MAE, Quantile, LogLinQuantile, MAPE, SMAPE, Poisson, MultiClass, MultiClassOneVsAll, PairLogit, QueryRMSE, QuerySoftmax. A loss might have params, then params should be written in format Loss:paramName=value.")
```
Comment (annaveronika) [inline catboost/app/bind_options.cpp]: Accidentally changed QuerySotfmax to QuerySoftMax

**Comment by lakshayg** [comment]
@annaveronika  I get this when I try to run the test.

```
➜  pytest git:(smape) ../../ya make -t -A
------- [SV] $(B)/library/svnversion/svnversion_data.h
Can't handle args: svn info --buffered /home/lakshayg/projects/catboost

Running LANG=C /home/lakshayg/.ya/tools/v3/481823646/ymake --python /home/lakshayg/projects/catboost/ya -v --no-report svn info --buffered /home/lakshayg/projects/catboost failed with exit code 1
|43.4%| PREPARE $(GDB)  142.3s
```

The progress gets stuck on 43.4%

File: catboost/libs/metrics/metric.cpp [hunk 4]
```diff
@@ -279,6 +279,39 @@ bool TMAPEMetric::IsMaxOptimal() const {
...
+
+    return error;
+}
+
+TString TSMAPEMetric::GetDescription() const {
+    return ToString(ELossFunction::SMAPE);
+}
+
+bool TSMAPEMetric::IsMaxOptimal() const {
+    return true;
```
Comment (annaveronika) [inline catboost/libs/metrics/metric.cpp]: should this be false?

Comment (lakshayg) [reply catboost/libs/metrics/metric.cpp]: What I inferred from IsMaxOptimal is that the metric is bounded by finite values (Is my inference correct?). Since SMAPE is bounded on both ends I set this to true

Comment (annaveronika) [reply catboost/libs/metrics/metric.cpp]: This means - if metric best possible value is as large as possible. This is used for overfitting detection, selecting best iteration with use_best_model and in visualization when drawing best point on graph.

**Comment by annaveronika** [comment]
@lakshayg Has the process finished? According to the error, the gdb is downloading slow, maybe you have slow network.

**Comment by lakshayg** [comment]
Yes, the process finished.

File: catboost/libs/metrics/metric.cpp (hunk 4 above)
Comment (lakshayg) [reply catboost/libs/metrics/metric.cpp]: Oh, that means I interpreted incorrectly. Let me fix that.

**Comment by annaveronika** [comment]
@lakshayg ping

**Comment by lakshayg** [comment]
Hi. I haven't forgotten this :)

I am still trying to debug the issue. Will update when I make some progress. (I'm able to work on this only on weekends)

**Comment by belonesox** [comment]
What is the state of SMAPE support?
* https://tech.yandex.com/catboost/doc/dg/concepts/loss-functions-docpage/#loss-functions__regression ← ``SMAPE`` metric declared as implemeted.  
* But 10.4.1 still fails with "Unknown loss function SMAPE"

?

**Comment by annaveronika** [comment]
Sorry, we'll fix the docs, it's only implemented as a metric to calculate, not to optimize now.
@lakshayg Do you have a plan to implement it as an optimizable loss function?

**Comment by lakshayg** [comment]
Sorry for the late update. I haven't been able to find time to work on this issue. I would be happy if someone willing to take over this issue.

**Comment by dimka11** [comment]
@lakshayg any updates?
